src/scene/shader-lib/shader-utils.js

```diff
@@ -0,0 +1,191 @@
+                fragmentDefines: options.fragmentDefines,
+                fragmentOutputTypes: options.fragmentOutputTypes
+            }));
+            programLibrary.setCachedShader(options.uniqueName, shader);
+        }
+        return shader;
+    }
+
+    /**
+     * Create a map of defines used for shader generation for a material.
+     *
+     * @param {Material} material - The material to create the shader defines for.
+     * @param {ShaderVariantParams} params - The shader variant parameters.
+     * @returns {Map<string, string>} The map of shader defines.
+     * @ignore
+     */
+    static getCoreDefines(material, params) {
+
+        // merge both maps, with camera shader params taking precedence
+        const defines = new Map(material.defines);
+        params.cameraShaderParams.defines.forEach((value, key) => defines.set(key, value));
+
+        // add pass defines
+        const shaderPassInfo = ShaderPass.get(params.device).getByIndex(params.pass);
+        shaderPassInfo.defines.forEach((value, key) => defines.set(key, value));
+
+        return defines;
+    }
+
+    /**
+     * Process shader using shader processing options, utilizing cache of the ProgramLibrary
+     *
+     * @param {Shader} shader - The shader to be processed.
+     * @param {ShaderProcessorOptions} processingOptions - The shader processing options.
+     * @returns {Shader} The processed shader.
+     * @ignore
+     */
+    static processShader(shader, processingOptions) {
+
+        Debug.assert(shader);
+        const shaderDefinition = shader.definition;
+
+        // 'shader' generator for a material - simply return existing shader definition. Use generator and getProgram
+        // to allow for shader processing to be cached
+        const name = shaderDefinition.name ?? 'shader';
+
+        // unique name based of the shader id
+        const key = `${name}-id-${shader.id}`;
+
+        const materialGenerator = new ShaderGeneratorPassThrough(key, shaderDefinition);
+
+        // temporarily register the program generator
+        const libraryModuleName = 'shader';
+        const library = getProgramLibrary(shader.device);
+        Debug.assert(!library.isRegistered(libraryModuleName));
+        library.register(libraryModuleName, materialGenerator);
+
+        // generate shader variant - its the same shader, but with different processing options
+        const variant = library.getProgram(libraryModuleName, {}, processingOptions);
+
+        // unregister it again
+        library.unregister(libraryModuleName);
+
+        return variant;
+    }
+}
+
+function createShader(device, vsName, fsName, useTransformFeedback = false, shaderDefinitionOptions = {}) {
+    Debug.removed('pc.createShader has been removed deprecated. Use ShaderUtils.createShader instead.');
+}
+
+function createShaderFromCode(device, vsCode, fsCode, uniqueName, attributes, useTransformFeedback = false, shaderDefinitionOptions = {}) {
```

Comment:
shouldn't this live in deprecated?

Reply:
this is the current thinking, to improve three-shaking:
https://github.com/playcanvas/engine/pull/6900